assets/migrations/mysql/001_initialize_schema.sql

```diff
@@ -0,0 +1,58 @@
+-- +goose Up
+CREATE TABLE tuple (
+		store VARCHAR(128) NOT NULL,
```

Comment:
In MySQL, we cannot use `TEXT` in keys as they keys would be too big (max ~3000 bytes), so we have to limit the length of each of the elements.
Have to double check if these are applicable bounds or if I should make another primary key and use indexes instead to increase the allowed length of the entries.

Reply:
Let's see:
- Store IDs are ulids so have a max of 26 characters.
- The total number of characters in an object = objectType + ":" + objectID is limited to 256
- Relations are limited to 50 characters
- Users are limited to 512 characters

You can mostly see those limits [here](https://github.com/openfga/api/blob/53441305ac6b0bc71602d6b2e918846047df37e8/openfga/v1/openfga.proto#L34).

Therefore, I think you can do something like:
```
CREATE TABLE tuple (
		store CHAR(26) NOT NULL,
		object_type VARCHAR(256) NOT NULL,
		object_id VARCHAR(256) NOT NULL,
		relation VARCHAR(50) NOT NULL,
		_user VARCHAR(512) NOT NULL,
		user_type VARCHAR(7) NOT NULL,
		ulid CHAR(26) NOT NULL,
		inserted_at TIMESTAMP NOT NULL,
		PRIMARY KEY (store, object_type, object_id, relation, _user)
);
```
and you will be safe. (I read that CHAR is better when you know the length.)

Reply:
Unfortunately that's still too long for having it as primary key. A better but still not the same limit as you have is this: 
```sql
CREATE TABLE tuple (
		store CHAR(26) NOT NULL,
		object_type VARCHAR(200) NOT NULL,
		object_id VARCHAR(200) NOT NULL,
		relation VARCHAR(50) NOT NULL,
		_user VARCHAR(256) NOT NULL,
		user_type VARCHAR(7) NOT NULL,
		ulid CHAR(26) NOT NULL,
		inserted_at TIMESTAMP NOT NULL,
		PRIMARY KEY (store, object_type, object_id, relation, _user)
);
```
I could also set a "virtual" primary key (just an autoincremented number), and add indices on all the columns that compose the primary key in the postgres engine.

Reply:
Since I can't create a unique constraint (it's too big), I've tried to resolve it in code, but I'm not 100% sure the solution is correct. I think I'll have to make use of [gap locks](https://dev.mysql.com/doc/refman/8.0/en/innodb-locking.html#innodb-gap-locks) to prevent something from inserting a record after the select statement has checked for uniqueness.
But to use the gap lock I would need to generate a "min" and "max" ulid that I would want to lock between, which seems like a rather odd solution, as the min and max would also be sort of random

Reply:
Oh my. I'd my rather avoid that. In the future I want to refactor the Postgres code to use `database/sql` and reduce the duplication.

Okay, if I understand correctly, MySQL with the InnoDB engine, can handle primary keys upto 1024 bytes. Or do we need it to be under 500 bytes? In any case, let's just fit in there but cutting down stuff. WDYT? If someone needs a bigger object_type or something we can think about that later.
```
CREATE TABLE tuple (
		store CHAR(26) NOT NULL,
		object_type VARCHAR(100) NOT NULL,
		object_id VARCHAR(100) NOT NULL,
		relation VARCHAR(50) NOT NULL,
		_user VARCHAR(200) NOT NULL,
		user_type VARCHAR(7) NOT NULL,
		ulid CHAR(26) NOT NULL,
		inserted_at TIMESTAMP NOT NULL,
		PRIMARY KEY (store, object_type, object_id, relation, _user)
);
```
Does that work, or should we cut even more?

Reply:
I did a `brew install mysql`, and tried this out from the command line and it seemed to work 🎉

Reply:
If I'm not mistaken, it's a 1024 byte limit per column, and for a primary key over multiple columns, the max is 3072 bytes (which doesn't coincide with # of characters). If the lower limits are fine I'll play around a bit more to see if we increase it a bit more and remove all the manual uniqueness verification!

Reply:
Great! I think that is fine for now. Would you mind documenting those limits in your [docs PR](https://github.com/openfga/openfga.dev/pull/198), and also in the README here? I think you can just mention that the limits are slightly different and link to the docs.

Thanks so much again for the contribution! It looks great.

Reply:
I've added the limitations in the docs pr as well as a reference to them in the README here. I'm not sure if they are in an appropriate spot in the docs, but I couldn't find a more appropriate spot to talk about the limitations of the datastore engine.

Reply:
Thanks! LGTM.